Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs

```diff
@@ -0,0 +1,137 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Unity.Netcode;
+using UnityEditor;
+using UnityEditor.SceneManagement;
+using UnityEngine;
+
+namespace Unity.Multiplayer.Samples.Utilities
+{
+    /// <summary>
+    /// Custom spawning component to be added to a scene GameObject. This component collects NetworkObjects in a scene
+    /// marked by a special tag, collects their Transform data, destroys their prefab instance, and performs the dynamic
+    /// spawning of said objects during Netcode for GameObject's (Netcode) OnNetworkSpawn() callback.
+    /// </summary>
+    [RequireComponent(typeof(NetcodeHooks))]
+    public class NetworkObjectSpawner : MonoBehaviour
+    {
+        [SerializeField]
+        NetcodeHooks m_NetcodeHooks;
+
+        [SerializeField]
+        SpawnObjectData m_SpawnObjectDataPrefab;
+
+        [SerializeField]
+        List<SpawnObjectData> m_SpawnObjectData;
+
+        const string k_NetworkObjectSpawnerCollectableTag = "NetworkObjectSpawnerCollectable";
+
+        void Awake()
+        {
+            m_NetcodeHooks.OnNetworkSpawnHook += OnNetworkSpawn;
+        }
+
+        void OnDestroy()
+        {
+            if (m_NetcodeHooks)
+            {
+                m_NetcodeHooks.OnNetworkSpawnHook -= OnNetworkSpawn;
+            }
+        }
+
+        void OnNetworkSpawn()
+        {
+            if (!NetworkManager.Singleton.IsServer)
+            {
+                enabled = false;
+                return;
+            }
+
+            StartCoroutine(WaitToSpawnNetworkObjects());
+        }
+
+        IEnumerator WaitToSpawnNetworkObjects()
+        {
+            // must wait for Netcode's OnNetworkSpawn() sweep before dynamically spawning
+            yield return new WaitForEndOfFrame();
+            SpawnNetworkObjects();
+        }
+
+        void SpawnNetworkObjects()
+        {
+            for (int i = m_SpawnObjectData.Count - 1; i >= 0; i--)
+            {
+                var spawnedGameObject = Instantiate(m_SpawnObjectData[i].prefabReference,
+                    m_SpawnObjectData[i].transform.position,
+                    m_SpawnObjectData[i].transform.rotation,
+                    null);
+
+                spawnedGameObject.transform.localScale = m_SpawnObjectData[i].transform.lossyScale;
+                var spawnedNetworkObject = spawnedGameObject.GetComponent<NetworkObject>();
+                spawnedNetworkObject.Spawn();
+
+                Destroy(m_SpawnObjectData[i].gameObject);
+            }
+        }
+
+#if UNITY_EDITOR
+        public void CollectTaggedPrefabInstances()
```

Comment:
I'm confused, didn't we say we'd have one spawner instance per static instance? And so we'd have 100 spawners in scene, for each placed imp? And this way, each spawner could reference the imp prefab to spawn and no need for all this logic to maintain?

Reply:
This is the spawner object you'd ideally drop into your scene that already contained NetworkObjects, and there'd be only one of these ideally. We don't have spawner instances, but just spawner data GameObjects, such that we're still able to move/rotate transforms after converting them into data GameObjects.

Reply:
we'd still have all this extra logic to maintain no? why the complexity?

Reply:
I kept this around initially to help myself convert the prefab instances into the data gameobjects, but if we don't wish to keep this around, we can enforce dropping the data NetworkObject prefab into scenes as the only flow for spawning in-scene placed networkobjects.

Reply:
the less code we need to maintain and debug, the better.
the end goal is to have something that's easy to use by a game designer. the simpler the better!